In [1]:
import os
import pandas as pd
from dotenv import load_dotenv
from utils.coin import CoinGeckoAPI  
from utils.manager import BitcoinDatabase  

load_dotenv()

db_name = os.getenv("BTC_DB")
table_btc_price_history = os.getenv("BTC_PRICE_TABLE")
table_rolling_price_history = os.getenv("ROLLING_PRICE_TABLE")



c:\users\toglia\appdata\local\programs\python\python38\lib\site-packages\pandas\core\computation\expressions.py:20: UserWarning: Pandas requires version '2.7.3' or newer of 'numexpr' (version '2.7.1' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED


##  Load the data from the db 

In [21]:
btc_db = BitcoinDatabase(db_name,table_btc_price_history)
btc_history_data = btc_db.select_data(table_btc_price_history)



## Find peak values calculating std

If a value is greater or less than two times the standard deviation, it can be considered an anomaly 

In [22]:
df = btc_history_data.copy()

In [23]:
df.describe()

price
count   2161.000000
mean   41286.376671
std     3085.657626
min    33505.713900
25%    38799.134300
50%    41702.562600
75%    43383.780100
max    48033.245200

In [24]:
desviacion_estandar = df['price'].std()

# Encontrar los valores que están por encima o por debajo de 2 desviaciones estándar
valores_anomalos = df[(df['price'] > (df['price'].mean() + 2 * desviacion_estandar)) | 
                      (df['price'] < (df['price'].mean() - 2 * desviacion_estandar))]

# Etiquetar los valores anómalos en el DataFrame original
df['anomalo'] = False
df.loc[valores_anomalos.index, 'anomalo'] = True




## Anomaly values

In [25]:
df[df['anomalo']==True]

date       price  anomalo
14    01-01-2022 14:06:31  48033.2452     True
15    01-01-2022 15:03:26  47648.6577     True
17    01-01-2022 17:06:07  47490.0659     True
19    01-01-2022 19:04:22  47506.7727     True
20    01-01-2022 20:01:18  47816.0777     True
21    01-01-2022 21:04:15  47692.3364     True
23    01-01-2022 23:00:06  47482.1361     True
37    02-01-2022 13:02:39  47861.4554     True
56    03-01-2022 08:00:33  47503.7562     True
83    04-01-2022 11:01:44  47507.6937     True
510   22-01-2022 06:00:02  34658.6912     True
516   22-01-2022 12:00:00  34991.0205     True
517   22-01-2022 13:01:25  35109.6565     True
518   22-01-2022 14:00:29  34631.7310     True
519   22-01-2022 15:07:59  34527.6528     True
520   22-01-2022 16:01:47  34602.7950     True
523   22-01-2022 19:02:32  34935.3106     True
527   22-01-2022 23:00:30  35044.5926     True
543   23-01-2022 15:01:27  35005.5795     True
544   23-01-2022 16:03:20  34804.5607     True
556   24-01-2022 04:03:07  34933.3911     True
557   24-01-2022 05:01:08  35059.2342     True
558   24-01-2022 06:03:02  34484.2537     True
559   24-01-2022 07:03:18  33505.7139     True
560   24-01-2022 08:02:52  33795.3241     True
561   24-01-2022 09:02:04  33520.3357     True
562   24-01-2022 10:02:18  33556.5981     True
563   24-01-2022 11:03:20  33869.7338     True
564   24-01-2022 12:00:17  34469.5473     True
565   24-01-2022 13:03:19  34455.4312     True
1295  24-02-2022 02:01:50  34740.0119     True
1296  24-02-2022 03:03:51  34987.4174     True
2074  28-03-2022 10:00:06  47600.8482     True
2075  28-03-2022 11:00:38  47557.1147     True
2078  28-03-2022 14:03:25  47555.5915     True
2079  28-03-2022 15:01:05  47836.9228     True
2080  28-03-2022 16:00:51  47785.1995     True
2081  28-03-2022 17:00:20  47878.5897     True
2082  28-03-2022 18:02:17  47603.8337     True
2083  28-03-2022 19:01:58  47560.5944     True
2086  28-03-2022 22:02:30  47513.0543     True
2089  29-03-2022 01:00:07  47516.0804     True
2090  29-03-2022 02:01:48  47594.1597     True
2091  29-03-2022 03:02:08  47509.9641     True
2093  29-03-2022 05:01:24  47497.7564     True
2094  29-03-2022 06:02:10  47550.8861     True
2095  29-03-2022 07:10:00  47562.1946     True
2096  29-03-2022 08:00:44  47884.6928     True
2097  29-03-2022 09:02:30  47825.8649     True
2098  29-03-2022 10:02:05  47652.0469     True
2099  29-03-2022 11:02:06  47807.3038     True
2100  29-03-2022 12:01:37  47503.5473     True
2103  29-03-2022 15:00:39  47522.0422     True
2104  29-03-2022 16:02:11  47667.1695     True
2127  30-03-2022 14:02:13  47651.0004     True

In [29]:
# Processing date to obtain more info

df['date'] = pd.to_datetime(df['date'], format='%d-%m-%Y %H:%M:%S')
df['day'] = df['date'].apply(lambda x: x.day)
df['month'] = df['date'].apply(lambda x: x.month)
df['week'] = df['date'].apply(lambda x: x.isocalendar()[1])

In [30]:
# Agrupar los datos por día, mes y semana y calcular la variación del precio de Bitcoin
meses_variacion = df.groupby('month')['price'].std().sort_values(ascending=False)
semanas_variacion = df.groupby('week')['price'].std().sort_values(ascending=False)



print("Semanas con mayor variación en el precio de Bitcoin:")
print(semanas_variacion.to_frame()[0:3] )

print("Mes con mayor variación en el precio de Bitcoin:")
print(meses_variacion.to_frame())

Semanas con mayor variación en el precio de Bitcoin:
            price
week             
3     3044.282615
9     2216.861163
1     2179.998169
Mes con mayor variación en el precio de Bitcoin:
             price
month             
1      3623.356522
3      2815.652961
2      2535.325682


## Calculating the IQR 

In [31]:
# Calcular el primer y tercer cuartil
q1 = df['price'].quantile(0.25)
q3 = df['price'].quantile(0.75)

# Calcular el rango intercuartil
iqr = q3 - q1

# Imprimir el rango intercuartil
print("El rango intercuartil es:", iqr)

El rango intercuartil es: 4584.645800000006


## According the IQR obtained the 50% of the values  have a spread of  4584.6458$